### DISCUSSION TITLE
Minimal container registry

### DISCUSSION THREAD

**Discussion by mwopfner** [containers/podman#17711] [Q&A]
Hi, I'm looking for a small lib (with little external dependencies) to spin up a local container registry from my own golang CLI tool. 
It is only used for development and testing. 

I found zot (https://github.com/project-zot/zot) but this has a huge dependency list. Because I already use `containers/image` I was wondering myself if there are such functions in the `containers/*` repos which are used to spinup a small registry for testing.

**Comment by afbjorklund** [comment]
https://docs.docker.com/registry/

https://github.com/containers/podman/blob/main/hack/podman-registry

```console
$ podman-registry start
PODMAN_REGISTRY_IMAGE="docker.io/library/registry:2.8"
PODMAN_REGISTRY_PORT="5050"
PODMAN_REGISTRY_USER="userZ3RZ"
PODMAN_REGISTRY_PASS="T8JVJzKrcl4p6uT"
```

https://github.com/distribution/distribution

**Reply by mwopfner** [reply]
@afbjorklund Thanks for the links, but they are also rather big for my purpose. 

I only need a registry which allows `pull` and I hoped there is something hidden in the `containers/*` repos which is used there for testing.

**Reply by afbjorklund** [reply]
If you find one, it would be interesting also for containerd and others!

I only know of even **bigger** projects, such as CNCF Harbor, that include registry/distribution as one of the dependencies...

https://www.docker.com/blog/donating-docker-distribution-to-the-cncf/

But maybe I will go with something like `nerdctl ipfs registry`, and then see if podman comes up with something new ?

**Reply by afbjorklund** [reply]
I guess one should also mention quay, I think that is the party line.

https://github.com/quay/quay

**Reply by mwopfner** [reply]
Yes bigger projects are easy to find :wink:
I actually only want a golang lib which allows me to start a simple registry from my own cli tool.

**Reply by afbjorklund** [reply]
It's mostly a glorified webserver, so with the right framework I _think_ it should be rather small ?

**Reply by mwopfner** [reply]
Yes, that's correct. It's not much, that is why I hoped there is a small implementation somewhere where I can start from or which I could reuse. If there is nothing I probably have to write it by myself.

**Reply by afbjorklund** [reply]
A small  and portable registry server would be useful to have, it could be used to cache images on the host for "machine".

Currently we are resorting to ugly measures such as `save` and `load`, but lose all the layer niceties from `push` and `pull`

**Reply by afbjorklund** [reply]
The docker registry is not _enormous_ (24M), but I'm not sure it would work as a standalone program on e.g. Mac or Win ? 

```
REPOSITORY   TAG       IMAGE ID       CREATED      SIZE
registry     2         8db46f9d7550   7 days ago   24.2MB
```

Some kind of PoC would be interesting, maybe I will hack something together... This looks neat: https://echo.labstack.com/
